homeassistant/components/octoprint/sensor.py

```diff
@@ -1,139 +1,234 @@
+        entities.append(OctoPrintEstimatedFinishTimeSensor(coordinator, name))
+    if "Time Elapsed" in monitored_conditions:
+        entities.append(OctoPrintStartTimeSensor(coordinator, name))
+
+    async_add_entities(entities, True)
+
+
+class OctoPrintSensorBase(CoordinatorEntity, Entity):
     """Representation of an OctoPrint sensor."""
 
     def __init__(
         self,
-        api,
-        condition,
-        sensor_type,
-        sensor_name,
-        unit,
-        endpoint,
-        group,
-        tool=None,
-        icon=None,
+        coordinator: DataUpdateCoordinator,
+        sensor_name: str,
+        sensor_type: str,
     ):
         """Initialize a new OctoPrint sensor."""
-        self.sensor_name = sensor_name
-        if tool is None:
-            self._name = f"{sensor_name} {condition}"
-        else:
-            self._name = f"{sensor_name} {condition} {tool} temp"
-        self.sensor_type = sensor_type
-        self.api = api
+        super().__init__(coordinator)
         self._state = None
-        self._unit_of_measurement = unit
-        self.api_endpoint = endpoint
-        self.api_group = group
-        self.api_tool = tool
-        self._icon = icon
-        _LOGGER.debug("Created OctoPrint sensor %r", self)
+        self._sensor_name = sensor_name
+        self._name = f"{sensor_name} {sensor_type}"
 
     @property
     def name(self):
         """Return the name of the sensor."""
         return self._name
 
+
+class OctoPrintStatusSensor(OctoPrintSensorBase):
+    """Representation of an OctoPrint sensor."""
+
+    def __init__(self, coordinator: DataUpdateCoordinator, sensor_name: str):
+        """Initialize a new OctoPrint sensor."""
+        super().__init__(coordinator, sensor_name, "Current State")
+
+    @property
+    def state(self):
+        """Return sensor state."""
+        printer: OctoprintPrinterInfo = self.coordinator.data["printer"]
+        if not printer:
+            return None
+
+        return printer.state.text
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend."""
+        return "mdi:printer-3d"
+
+
+class OctoPrintJobPercentageSensor(OctoPrintSensorBase):
+    """Representation of an OctoPrint sensor."""
+
+    def __init__(self, coordinator: DataUpdateCoordinator, sensor_name: str):
+        """Initialize a new OctoPrint sensor."""
+        super().__init__(coordinator, sensor_name, "Job Percentage")
+
     @property
     def state(self):
-        """Return the state of the sensor."""
-        sensor_unit = self.unit_of_measurement
-        if sensor_unit in (TEMP_CELSIUS, PERCENTAGE):
-            # API sometimes returns null and not 0
-            if self._state is None:
-                self._state = 0
-            return round(self._state, 2)
-        return self._state
+        """Return sensor state."""
+        job: OctoprintJobInfo = self.coordinator.data["job"]
+        if not job:
+            return 0
```

Comment:
If there is no job, shouldn't this be unknown? (and thus return `None`)

Reply:
Was the existing behavior, I will swap that over